File: app/src/main/java/com/example/locationmocker/domain/route/PlaybackCursor.kt
package com.example.locationmocker.domain.route

import com.example.locationmocker.domain.model.PlaybackMode
import com.example.locationmocker.domain.model.RouteSample

class PlaybackCursor(
    private val samples: List<RouteSample>,
    private val mode: PlaybackMode,
) {
    private var index = 0
    private var direction = 1
    private var completed = false

    fun next(): RouteSample? {
        if (samples.isEmpty() || completed) return null
        if (samples.size == 1) {
            if (mode == PlaybackMode.Once) completed = true
            return samples.first()
        }

        val current = samples[index]
        advance()
        return current
    }

    private fun advance() {
        when (mode) {
            PlaybackMode.Once -> {
                if (index >= samples.lastIndex) {
                    completed = true
                } else {
                    index++
                }
            }

            PlaybackMode.Loop -> {
                index = if (index >= samples.lastIndex) 0 else index + 1
            }

            PlaybackMode.PingPong -> {
                val nextIndex = index + direction
                when {
                    nextIndex > samples.lastIndex -> {
                        direction = -1
                        index = samples.lastIndex - 1
                    }

                    nextIndex < 0 -> {
                        direction = 1
                        index = 1
                    }

                    else -> index = nextIndex
                }
            }
        }
    }
}

File: ios/LocationMocker/LocationMocker/Domain/Route/PlaybackCursor.swift
import Foundation

final class PlaybackCursor {
    private let samples: [RouteSample]
    private let mode: PlaybackMode
    private var index = 0
    private var direction = 1
    private var completed = false

    init(samples: [RouteSample], mode: PlaybackMode) {
        self.samples = samples
        self.mode = mode
    }

    func next() -> RouteSample? {
        guard !samples.isEmpty, !completed else { return nil }
        if samples.count == 1 {
            if mode == .once { completed = true }
            return samples[0]
        }
        let current = samples[index]
        advance()
        return current
    }

    func reset() {
        index = 0
        direction = 1
        completed = false
    }

    private func advance() {
        switch mode {
        case .once:
            if index >= samples.count - 1 {
                completed = true
            } else {
                index += 1
            }
        case .loop:
            index = index >= samples.count - 1 ? 0 : index + 1
        case .pingPong:
            let nextIndex = index + direction
            if nextIndex > samples.count - 1 {
                direction = -1
                index = samples.count - 2
            } else if nextIndex < 0 {
                direction = 1
                index = 1
            } else {
                index = nextIndex
            }
        }
    }
}
